redis/src/sentinel.rs

```diff
@@ -0,0 +1,558 @@
+    }
+
+    /// Determines the masters address for the given name, and returns a client for that
+    /// master.
+    ///
+    /// To determine the master we ask the sentinels one by one until one of
+    /// then correctly responds to the SENTINEL MASTERS command, and then we check to
+    /// see if there is a master with the given name; if there is, we create a
+    /// client with that address (ip and port).
+    pub async fn async_master_for(&self, service_name: &str) -> RedisResult<Client> {
+        let address = self.async_find_master_address(service_name).await?;
+        Client::open(address)
+    }
+
+    /// Connects to the first available replica of the given master name.
+    ///
+    /// After listing the valid replicas (those without a sdown or odown flag) for the
+    /// given master name, we start trying to connect to then one by one, starting at a
+    /// random index (simple random derived from the nanoseconds part of the current
+    /// unix timestamp).
+    pub async fn async_replica_for(&self, service_name: &str) -> RedisResult<Client> {
+        let addresses = self
+            .async_find_valid_replica_addresses(service_name)
+            .await?;
+        let start_index = random_replica_index(addresses.len());
+        try_connect_to_first_replica(&addresses, start_index)
+    }
+
+    /// Connects to the first available replica of the given master name, starting in a
+    /// different (incremented) index each time the function is called.
+    ///
+    /// After listing the valid replicas (those without a sdown or odown flag) for the
+    /// given master name, we start trying to connect to then one by one, starting at a
+    /// rotating index which gets incremented on each call to this function.
+    pub async fn async_replica_rotate_for(&mut self, service_name: &str) -> RedisResult<Client> {
+        let addresses = self
+            .async_find_valid_replica_addresses(service_name)
+            .await?;
+        if !addresses.is_empty() {
+            self.replica_start_index = (self.replica_start_index + 1) % addresses.len();
+        }
+        try_connect_to_first_replica(&addresses, self.replica_start_index)
+    }
+}
+
+// non-async methods
+impl Sentinel {
+    /// Creates a Sentinel client performing some basic
+    /// checks on the URLs that might make the operation fail.
+    pub fn build<T: IntoConnectionInfo>(params: Vec<T>) -> RedisResult<Sentinel> {
+        Ok(Sentinel {
+            sentinels_connection_info: params
+                .into_iter()
+                .map(|p| p.into_connection_info())
+                .collect::<RedisResult<Vec<ConnectionInfo>>>()?,
+            replica_start_index: random_replica_index(1000000),
+        })
+    }
+
+    fn try_all_sentinels<F, T>(&self, f: F) -> RedisResult<T>
+    where
+        F: Fn(&ConnectionInfo) -> RedisResult<T>,
+    {
+        let mut last_err = None;
+        for connection_info in self.sentinels_connection_info.iter() {
```

Comment:
To do that, I'd have to require a mutable reference, or use interior mutability, right? What's the most common/recommended approach in this case?

Reply:
interesting question.
In general I would've gone with interior mutability if you believe this method might be called in parallel from multiple sources, and mutable reference if you believe that this method usually would be called once every so often.
If you're going with interior mutability, make sure that you're using a mutex that is safe to hold over an await point, since the mutex might be used in both sync and async methods.